## Nomor 1

In [2]:
import numpy as np # Import numpy untuk array
from tabulate import tabulate # Import tabulate untuk tabel

aug_matrix = np.array(eval(input('Masukkan augmented matrix dari SPL yang akan diselesaikan: '))) # Simpan matriks yang akan diselesaikan
initial_solution = np.array(eval(input('Masukkan tebakan awal solusi: ')),dtype=float) # Simpan tebakan awal solusi dari user (x0)
tolerance = eval(input('Masukkan toleransi: ')) # Simpan batas toleransi yang diinginkan

print(aug_matrix) # Cetak matriks

def Jacobi(matrix,initial,tol): # Buat fungsi penyelesaian SPL dengan metode Jacobi
    n = np.shape(matrix)[0] # Ukuran matriks/banyak variabel
    x0 = initial # Definisikan x0 sebagai tebakan awal
    x = x0.copy() # Definisikan x = x0 (gunakan .copy() agar jika x0 berubah, x tidak berubah)
    
    headers=["Iterasi"] # Buat list utk headers tabel
    
    for i in range(n):
        headers.append("x"+str(i+1))
        b = matrix[i,n] # Konstanta pada SPL
        for j in range(n): # Metode Jacobi
            if j!=i:
                b-=matrix[i,j]*x0[j] 
        x[i]=b/matrix[i,i] # Hasil aproksimasi Jacobi
    
    list = [] # Inisiasi bentuk list yang akan diisi ke tabel
    iterasi = 1 # Inisiasi nilai jumlah iterasi
    solx = np.insert(x,0,iterasi) # Masukkan list x yang diiterasi ke-sekian pada list
    list.append(solx) # Menambahkan list x ke list yang akan diisi ke tabel

    while np.linalg.norm(x-x0,np.inf)>=tol: # Pengecekan kriteria penyetop menggunakan error mutlak
        x0 = x # Definisikan x0 sebagai x(k-1)
        x = x0.copy() # Definisikan x sebagai x(k), namun awalnya samakan dengan x0
        for i in range(n): # Metode Jacobi
            b = matrix[i,n] # Konstanta pada SPL
            for j in range(n):
                if j!=i:
                    b-=matrix[i,j]*x0[j]
            x[i]=b/matrix[i,i] # Hasil aproksimasi metode Jacobi
        
        iterasi += 1 # Tambah iterasi sebanyak 1
        solx = np.insert(x,0,iterasi) # Masukkan list x yang diiterasi ke-sekian pada list
        list.append(solx) # Menambahkan list x ke list yang akan diisi ke tabel
        
    print(tabulate(list, headers, tablefmt = "fancy_grid")) # Cetak seluruh hasil x ke dalam bentuk tabel (per iterasi)
    return x # Nilai yang di-return fungsi ini adalah x(k)

print('Aproksimasi solusi SPL tersebut dengan metode Jacobi adalah:',Jacobi(aug_matrix,initial_solution,tolerance)) # Panggil fungsi Jacobi dengan parameter yang diisi

Masukkan augmented matrix dari SPL yang akan diselesaikan: [[3,-1,2,1],[1,2,1,4],[2,1,3,4]]
Masukkan tebakan awal solusi: [0,0,0]
Masukkan toleransi: 10**-5
[[ 3 -1  2  1]
 [ 1  2  1  4]
 [ 2  1  3  4]]
╒═══════════╤══════════╤═════════╤══════════╕
│   Iterasi │       x1 │      x2 │       x3 │
╞═══════════╪══════════╪═════════╪══════════╡
│         1 │ 0.333333 │ 2       │ 1.33333  │
├───────────┼──────────┼─────────┼──────────┤
│         2 │ 0.111111 │ 1.16667 │ 0.444444 │
├───────────┼──────────┼─────────┼──────────┤
│         3 │ 0.425926 │ 1.72222 │ 0.87037  │
├───────────┼──────────┼─────────┼──────────┤
│         4 │ 0.32716  │ 1.35185 │ 0.475309 │
├───────────┼──────────┼─────────┼──────────┤
│         5 │ 0.467078 │ 1.59877 │ 0.664609 │
├───────────┼──────────┼─────────┼──────────┤
│         6 │ 0.423182 │ 1.43416 │ 0.489026 │
├───────────┼──────────┼─────────┼──────────┤
│         7 │ 0.485368 │ 1.5439  │ 0.57316  │
├───────────┼──────────┼─────────┼──────────┤
│         8 │ 0

Program di atas adalah program untuk mencari aproksimasi SPL dengan metode Jacobi.

Idenya adalah mencari solusi SPL dari input matriks, tebakan awal, dan toleransi dari user dengan metode Jacobi serta menampilkan hasil tiap iterasi dan hasil akhirnya.

Algoritmanya:
1. *Import* module `numpy`, lalu panggil sebagai `np`
1. *Import* `tabulate` dari module `tabulate`
1. Simpan *input* matriks dari *user* dan simpan dalam variabel `aug_matrix`
1. Simpan *input* tebakan awal dari *user* dan simpan dalam variabel `initial_solution`
1. Simpan *input* batas toleransi dari *user* dan simpan dalam variabel `tolerance`
1. Cetak matriks `aug_matrix`
1. Buat fungsi metode Jacobi dengan *argument* `matrix`, `initial`, dan `tol`. Berdasarkan Modul 6 Praktikum Metnum, program metode Jacobi akan berjalan sesuai dengan ketentuan tersebut, kecuali:
    1. Buat list `headers` dengan isi string `"Iterasi"` setelah pendefinisian variabel
    1. Menambahkan variabel `x_(i+1)` di tiap iterasi ke dalam list `headers` pada for loop pertama
    1. Buat list kosong dengan nama `list`
    1. Inisialisasi jumlah iterasi dengan nilai `1` ke dalam variabel `iterasi`
    1. Masukkan list x yang sudah diiterasi ke dalam variabel `solx`
    1. Pada while loop:
        1. Tambah iterasi sebanyak 1
        1. Masukkan kembali list x yang sudah diiterasi ke dalam variabel `solx`
        1. Masukkan list x yang sudah diiterasi ke dalam variabel `solx`
    1. Cetak tabel hasil tiap iterasi dengan module `tabulate` dan modifikasi bentuk tabel agar lebih cantik
    1. *Return*-kan nilai `x` setelah fungsi selesai
1. Panggil fungsi Jacobi dengan argumen `aug_matrix`, `initial_solution`, dan `tolerance`. Kemudian, cetak hasil iterasi dari variabel `x` yang sudah di *return* dari fungsi.

In [1]:
import numpy as np # Import numpy untuk array
from tabulate import tabulate # Import tabulate untuk tabel

aug_matrix = np.array(eval(input('Masukkan augmented matrix dari SPL yang akan diselesaikan: '))) # Simpan matriks diperbesar yang akan diselesaikan
initial_solution = np.array(eval(input('Masukkan tebakan awal solusi: ')),dtype=float) # Simpan tebakan awal solusi dari user (x0)
tolerance = eval(input('Masukkan toleransi: ')) # Simpan batas toleransi yang diinginkan

print(aug_matrix) # Cetak matriks

def GaussSeidel(matrix,initial,tol): # Buat fungsi penyelesaian SPL dengan metode Gauss-Seidel
    n = np.shape(matrix)[0] # Ukuran matriks/banyak variabel
    x0 = initial # Definisikan x0 sebagai tebakan awal
    x = x0.copy() # Definisikan x = x0 (gunakan .copy() agar jika x0 berubah, x tidak berubah)

    headers = ["Iterasi"] # Buat list utk headers tabel
    
    for i in range(n):
        headers.append("x"+str(i+1))
        b = matrix[i,n] # Konstanta pada SPL
        for j in range(n): # Metode Gauss-Seidel
            if j!=i:
                if j<i:
                    b-=matrix[i,j]*x[j]
                else:
                    b-=matrix[i,j]*x0[j]
        x[i]=b/matrix[i,i] # Hasil aproksimasi Gauss-Seidel

    list = [] # Inisiasi bentuk list yang akan diisi ke tabel
    iterasi = 1 # Inisiasi nilai jumlah iterasi
    solx = np.insert(x,0,iterasi) # Masukkan list x yang diiterasi ke-sekian pada list
    list.append(solx) # Menambahkan list x ke list yang akan diisi ke tabel
    
    while np.linalg.norm(x-x0,np.inf)>=tol: # Pengecekan kriteria penyetop menggunakan error mutlak
        x0 = x # Definisikan x0 sebagai x(k-1)
        x = x0.copy() # Definisikan x sebagai x(k), namun awalnya samakan dengan x0
        for i in range(n): # Metode Gauss-Seidel
            b = matrix[i,n] # Konstanta pada SPL
            for j in range(n):
                if j!=i:
                    if j<i:
                        b-=matrix[i,j]*x[j]
                    else:
                        b-=matrix[i,j]*x0[j]
            x[i]=b/matrix[i,i] # Hasil aproksimasi metode Gauss-Seidel
        
        iterasi += 1 # Tambah iterasi sebanyak 1
        solx = np.insert(x,0,iterasi) # Masukkan list x yang diiterasi ke-sekian pada list
        list.append(solx) # Menambahkan list x ke list yang akan diisi ke tabel
        
    print(tabulate(list, headers, tablefmt = "fancy_grid")) # Cetak seluruh hasil x ke dalam bentuk tabel (per iterasi)
    return x # Nilai yang direturn fungsi ini adalah x(k)

print('Aproksimasi solusi SPL tersebut dengan metode Gauss-Seidel adalah:',GaussSeidel(aug_matrix,initial_solution,tolerance)) # Panggil fungsi Gauss-Seidel dengan parameter yang diisi

Masukkan augmented matrix dari SPL yang akan diselesaikan: [[12,7,3,2],[1,5,1,-5],[2,7,-11,6]]
Masukkan tebakan awal solusi: [1,3,5]
Masukkan toleransi: 10**-3
[[ 12   7   3   2]
 [  1   5   1  -5]
 [  2   7 -11   6]]
╒═══════════╤═══════════╤═══════════╤═══════════╕
│   Iterasi │        x1 │        x2 │        x3 │
╞═══════════╪═══════════╪═══════════╪═══════════╡
│         1 │ -2.83333  │ -1.43333  │ -1.97273  │
├───────────┼───────────┼───────────┼───────────┤
│         2 │  1.49596  │ -0.904646 │ -0.849146 │
├───────────┼───────────┼───────────┼───────────┤
│         3 │  0.906664 │ -1.0115   │ -1.02429  │
├───────────┼───────────┼───────────┼───────────┤
│         4 │  1.01278  │ -0.997698 │ -0.996211 │
├───────────┼───────────┼───────────┼───────────┤
│         5 │  0.99771  │ -1.0003   │ -1.00061  │
├───────────┼───────────┼───────────┼───────────┤
│         6 │  1.00033  │ -0.999944 │ -0.999905 │
├───────────┼───────────┼───────────┼───────────┤
│         7 │  0.999944 │ -1.000

Program di atas adalah program untuk mencari aproksimasi SPL dengan metode Gauss-Seidel.

Idenya adalah mencari solusi SPL dari input matriks, tebakan awal, dan toleransi dari user dengan metode Gauss-Seidel serta menampilkan hasil tiap iterasi dan hasil akhirnya.

Algoritmanya:
1. *Import* module `numpy`, lalu panggil sebagai `np`
1. *Import* `tabulate` dari module `tabulate`
1. Simpan *input* matriks dari *user* dan simpan dalam variabel `aug_matrix`
1. Simpan *input* tebakan awal dari *user* dan simpan dalam variabel `initial_solution`
1. Simpan *input* batas toleransi dari *user* dan simpan dalam variabel `tolerance`
1. Cetak matriks `aug_matrix`
1. Buat fungsi metode Gauss-Seidel dengan *argument* `matrix`, `initial`, dan `tol`. Berdasarkan Modul 6 Praktikum Metnum, program metode Gauss-Seidel akan berjalan sesuai dengan ketentuan tersebut, kecuali:
    1. Buat list `headers` dengan isi string `"Iterasi"` setelah pendefinisian variabel
    1. Menambahkan variabel `x_(i+1)` di tiap iterasi ke dalam list `headers` pada for loop pertama
    1. Buat list kosong dengan nama `list`
    1. Inisialisasi jumlah iterasi dengan nilai `1` ke dalam variabel `iterasi`
    1. Masukkan list x yang sudah diiterasi ke dalam variabel `solx`
    1. Pada while loop:
        1. Tambah iterasi sebanyak 1
        1. Masukkan kembali list x yang sudah diiterasi ke dalam variabel `solx`
        1. Masukkan list x yang sudah diiterasi ke dalam variabel `solx`
    1. Cetak tabel hasil tiap iterasi dengan module `tabulate` dan modifikasi bentuk tabel agar lebih cantik
    1. *Return*-kan nilai `x` setelah fungsi selesai
1. Panggil fungsi Gauss-Seidel dengan argumen `aug_matrix`, `initial_solution`, dan `tolerance`. Kemudian, cetak hasil iterasi dari variabel `x` yang sudah di *return* dari fungsi.

**Untuk $x^0=[1,1,1]$**

In [3]:
import numpy as np # Import numpy untuk array
from tabulate import tabulate # Import tabulate untuk tabel

aug_matrix = np.array(eval(input('Masukkan augmented matrix dari SPL yang akan diselesaikan: '))) # Simpan matriks yang akan diselesaikan
initial_solution = np.array(eval(input('Masukkan tebakan awal solusi: ')),dtype=float) # Simpan tebakan awal solusi dari user (x0)
tolerance = eval(input('Masukkan toleransi: ')) # Simpan batas toleransi yang diinginkan

print(aug_matrix) # Cetak matriks

def Jacobi(matrix,initial,tol): # Buat fungsi penyelesaian SPL dengan metode Jacobi
    n = np.shape(matrix)[0] # Ukuran matriks/banyak variabel
    x0 = initial # Definisikan x0 sebagai tebakan awal
    x = x0.copy() # Definisikan x = x0 (gunakan .copy() agar jika x0 berubah, x tidak berubah)
    
    headers=["Iterasi"] # Buat list utk headers tabel
    
    for i in range(n):
        headers.append("x"+str(i+1))
        b = matrix[i,n] # Konstanta pada SPL
        for j in range(n): # Metode Jacobi
            if j!=i:
                b-=matrix[i,j]*x0[j] 
        x[i]=b/matrix[i,i] # Hasil aproksimasi Jacobi
    
    list = [] # Inisiasi bentuk list yang akan diisi ke tabel
    iterasi = 1 # Inisiasi nilai jumlah iterasi
    solx = np.insert(x,0,iterasi) # Masukkan list x yang diiterasi ke-sekian pada list
    list.append(solx) # Menambahkan list x ke list yang akan diisi ke tabel

    while np.linalg.norm(x-x0,np.inf)>=tol: # Pengecekan kriteria penyetop menggunakan error mutlak
        x0 = x # Definisikan x0 sebagai x(k-1)
        x = x0.copy() # Definisikan x sebagai x(k), namun awalnya samakan dengan x0
        for i in range(n): # Metode Jacobi
            b = matrix[i,n] # Konstanta pada SPL
            for j in range(n):
                if j!=i:
                    b-=matrix[i,j]*x0[j]
            x[i]=b/matrix[i,i] # Hasil aproksimasi metode Jacobi
        
        iterasi += 1 # Tambah iterasi sebanyak 1
        solx = np.insert(x,0,iterasi) # Masukkan list x yang diiterasi ke-sekian pada list
        list.append(solx) # Menambahkan list x ke list yang akan diisi ke tabel
        
    print(tabulate(list, headers, tablefmt = "fancy_grid")) # Cetak seluruh hasil x ke dalam bentuk tabel (per iterasi)
    return x # Nilai yang di-return fungsi ini adalah x(k)

print('Aproksimasi solusi SPL tersebut dengan metode Jacobi adalah:',Jacobi(aug_matrix,initial_solution,tolerance)) # Panggil fungsi Jacobi dengan parameter yang diisi

Masukkan augmented matrix dari SPL yang akan diselesaikan: [[2,-1,1,-1],[2,2,2,4],[-1,-1,2,-5]]
Masukkan tebakan awal solusi: [1,1,1]
Masukkan toleransi: 10**-5
[[ 2 -1  1 -1]
 [ 2  2  2  4]
 [-1 -1  2 -5]]


C:\Users\Kamal\AppData\Local\Temp/ipykernel_4228/3018623884.py:37: RuntimeWarning: overflow encountered in double_scalars
  b-=matrix[i,j]*x0[j]
C:\Users\Kamal\AppData\Local\Temp/ipykernel_4228/3018623884.py:37: RuntimeWarning: invalid value encountered in double_scalars
  b-=matrix[i,j]*x0[j]
C:\Users\Kamal\AppData\Local\Temp/ipykernel_4228/3018623884.py:30: RuntimeWarning: invalid value encountered in subtract
  while np.linalg.norm(x-x0,np.inf)>=tol: # Pengecekan kriteria penyetop menggunakan error mutlak


╒═══════════╤════════════════════╤════════════════════╤════════════════════╕
│   Iterasi │                 x1 │                 x2 │                 x3 │
╞═══════════╪════════════════════╪════════════════════╪════════════════════╡
│         1 │      -0.5          │       0            │      -1.5          │
├───────────┼────────────────────┼────────────────────┼────────────────────┤
│         2 │       0.25         │       4            │      -2.75         │
├───────────┼────────────────────┼────────────────────┼────────────────────┤
│         3 │       2.875        │       4.5          │      -0.375        │
├───────────┼────────────────────┼────────────────────┼────────────────────┤
│         4 │       1.9375       │      -0.5          │       1.1875       │
├───────────┼────────────────────┼────────────────────┼────────────────────┤
│         5 │      -1.34375      │      -1.125        │      -1.78125      │
├───────────┼────────────────────┼────────────────────┼────────────────────┤

In [4]:
import numpy as np # Import numpy untuk array
from tabulate import tabulate # Import tabulate untuk tabel

aug_matrix = np.array(eval(input('Masukkan augmented matrix dari SPL yang akan diselesaikan: '))) # Simpan matriks diperbesar yang akan diselesaikan
initial_solution = np.array(eval(input('Masukkan tebakan awal solusi: ')),dtype=float) # Simpan tebakan awal solusi dari user (x0)
tolerance = eval(input('Masukkan toleransi: ')) # Simpan batas toleransi yang diinginkan

print(aug_matrix) # Cetak matriks

def GaussSeidel(matrix,initial,tol): # Buat fungsi penyelesaian SPL dengan metode Gauss-Seidel
    n = np.shape(matrix)[0] # Ukuran matriks/banyak variabel
    x0 = initial # Definisikan x0 sebagai tebakan awal
    x = x0.copy() # Definisikan x = x0 (gunakan .copy() agar jika x0 berubah, x tidak berubah)

    headers = ["Iterasi"] # Buat list utk headers tabel
    
    for i in range(n):
        headers.append("x"+str(i+1))
        b = matrix[i,n] # Konstanta pada SPL
        for j in range(n): # Metode Gauss-Seidel
            if j!=i:
                if j<i:
                    b-=matrix[i,j]*x[j]
                else:
                    b-=matrix[i,j]*x0[j]
        x[i]=b/matrix[i,i] # Hasil aproksimasi Gauss-Seidel

    list = [] # Inisiasi bentuk list yang akan diisi ke tabel
    iterasi = 1 # Inisiasi nilai jumlah iterasi
    solx = np.insert(x,0,iterasi) # Masukkan list x yang diiterasi ke-sekian pada list
    list.append(solx) # Menambahkan list x ke list yang akan diisi ke tabel
    
    while np.linalg.norm(x-x0,np.inf)>=tol: # Pengecekan kriteria penyetop menggunakan error mutlak
        x0 = x # Definisikan x0 sebagai x(k-1)
        x = x0.copy() # Definisikan x sebagai x(k), namun awalnya samakan dengan x0
        for i in range(n): # Metode Gauss-Seidel
            b = matrix[i,n] # Konstanta pada SPL
            for j in range(n):
                if j!=i:
                    if j<i:
                        b-=matrix[i,j]*x[j]
                    else:
                        b-=matrix[i,j]*x0[j]
            x[i]=b/matrix[i,i] # Hasil aproksimasi metode Gauss-Seidel
        
        iterasi += 1 # Tambah iterasi sebanyak 1
        solx = np.insert(x,0,iterasi) # Masukkan list x yang diiterasi ke-sekian pada list
        list.append(solx) # Menambahkan list x ke list yang akan diisi ke tabel
        
    print(tabulate(list, headers, tablefmt = "fancy_grid")) # Cetak seluruh hasil x ke dalam bentuk tabel (per iterasi)
    return x # Nilai yang direturn fungsi ini adalah x(k)

print('Aproksimasi solusi SPL tersebut dengan metode Gauss-Seidel adalah:',GaussSeidel(aug_matrix,initial_solution,tolerance)) # Panggil fungsi Gauss-Seidel dengan parameter yang diisi

Masukkan augmented matrix dari SPL yang akan diselesaikan: [[2,-1,1,-1],[2,2,2,4],[-1,-1,2,-5]]
Masukkan tebakan awal solusi: [1,1,1]
Masukkan toleransi: 10**-5
[[ 2 -1  1 -1]
 [ 2  2  2  4]
 [-1 -1  2 -5]]
╒═══════════╤═══════════╤═════════╤═══════════╕
│   Iterasi │        x1 │      x2 │        x3 │
╞═══════════╪═══════════╪═════════╪═══════════╡
│         1 │ -0.5      │ 1.5     │ -2        │
├───────────┼───────────┼─────────┼───────────┤
│         2 │  1.25     │ 2.75    │ -0.5      │
├───────────┼───────────┼─────────┼───────────┤
│         3 │  1.125    │ 1.375   │ -1.25     │
├───────────┼───────────┼─────────┼───────────┤
│         4 │  0.8125   │ 2.4375  │ -0.875    │
├───────────┼───────────┼─────────┼───────────┤
│         5 │  1.15625  │ 1.71875 │ -1.0625   │
├───────────┼───────────┼─────────┼───────────┤
│         6 │  0.890625 │ 2.17188 │ -0.96875  │
├───────────┼───────────┼─────────┼───────────┤
│         7 │  1.07031  │ 1.89844 │ -1.01562  │
├───────────┼───────────┼

### Jawaban
1. Diberikan SPL sebagai berikut: $$\begin{align*}2x_1-x_2+x_3&=-1,\\2x_1+2x_2+2x_3&=4,\\-x_1-x_2+2x_3&=-5\end{align*}$$ Dengan $x^0=0$, dan toleransi $10^-5$

    1. Metode Jacobi dan Gauss-Seidel untuk mengaproksimasi SPL dilampirkan di atas.

    1. Pada metode Jacobi, diperlukan sebanyak 6350 iterasi hingga memenuhi batas toleransi, sedangkan pada metode Gauss-Seidel, sebanyak 23 iterasi diperlukan hingga memenuhi	batas	toleransi.

    1. Iya, terdapat perbedaan jumlah iterasi antara metode Jacobi dan Gauss-Seidel. Pada metode Jacobi, terjadi iterasi sebanyak 6350, sedangkan pada metode Gauss-Seidel terjadi iterasi sebanyak 23.
    
    Hal ini membuktikan bahwa metode Gauss-Seidel lebih efektif untuk mempersingkat iterasi. Metode Gauss-Seidel lebih efektif karena terdapat penambahan kondisi dari metode Jacobi pada iterasi for loop. Metode Jacobi menggunakan nilai yang diperoleh dari langkah sebelumnya, sedangkan metode Gauss-Seidel selalu menggunakan nilai yang sudah diperbarui setiap iterasi.
    
    Selain itu, SPL di atas gagal diaproksimasi dengan metode Jacobi karena divergen. Hal ini terlihat pada iterasi ke-25, di mana nilainya semakin jauh dari nilai eksaknya ($[1,2,-1]$).

    1. Iya, didapatkan hasil hasil aproksimasi yang berbeda jika $x^0=[1,1,1]$.
    
        * Pada metode Jacobi, $x^0=[0,0,0]$ menghasilkan aproksimasi `[ inf  inf -inf]` dan $x^0=[1,1,1]$ menghasilkan `[-inf -inf  nan]`.
        
        * Pada metode Gauss-Seidel, $x^0=[0,0,0]$ menghasilkan aproksimasi `[ 1.00000226  1.9999975  -1.00000012]` dan $x^0=[1,1,1]$ menghasilkan `[ 0.99999744  2.0000028  -0.99999988]`.
        
        Terlihat bahwa banyaknya iterasi yang dilakukan juga berbeda, lebih banyak iterasi yang dilakukan jika $x^0=[1,1,1]$ daripada jika $x^0=[0,0,0]$. Hal ini terjadi karena $x^0=[0,0,0]$ lebih cepat konvergen ke nilai eksaknya ($[1,2,-1]$) dibandingkan dengan $x^0=[1,1,1]$, sudah terlihat pada iterasi pertamanya.

## Nomor 2

Pada nomor 2 ini, saya menghapus pembuatan tabel pada setiap iterasi agar program tetap memberikan output karena *overflow* dengan data yang sangat besar.

In [5]:
import numpy as np # Import numpy untuk array
from tabulate import tabulate # Import tabulate untuk tabel

aug_matrix = np.array(eval(input('Masukkan augmented matrix dari SPL yang akan diselesaikan: '))) # Simpan matriks yang akan diselesaikan
initial_solution = np.array(eval(input('Masukkan tebakan awal solusi: ')),dtype=float) # Simpan tebakan awal solusi dari user (x0)
tolerance = eval(input('Masukkan toleransi: ')) # Simpan batas toleransi yang diinginkan

print(aug_matrix) # Cetak matriks

def Jacobi(matrix,initial,tol): # Buat fungsi penyelesaian SPL dengan metode Jacobi
    n = np.shape(matrix)[0] # Ukuran matriks/banyak variabel
    x0 = initial # Definisikan x0 sebagai tebakan awal
    x = x0.copy() # Definisikan x = x0 (gunakan .copy() agar jika x0 berubah, x tidak berubah)
    
    for i in range(n):
        b = matrix[i,n] # Konstanta pada SPL
        for j in range(n): # Metode Jacobi
            if j!=i:
                b-=matrix[i,j]*x0[j] 
        x[i]=b/matrix[i,i] # Hasil aproksimasi Jacobi
    
    iterasi = 1 # Inisiasi nilai jumlah iterasi

    while np.linalg.norm(x-x0,np.inf)>=tol: # Pengecekan kriteria penyetop menggunakan error mutlak
        x0 = x # Definisikan x0 sebagai x(k-1)
        x = x0.copy() # Definisikan x sebagai x(k), namun awalnya samakan dengan x0
        for i in range(n): # Metode Jacobi
            b = matrix[i,n] # Konstanta pada SPL
            for j in range(n):
                if j!=i:
                    b-=matrix[i,j]*x0[j]
            x[i]=b/matrix[i,i] # Hasil aproksimasi metode Jacobi
        
        iterasi += 1 # Tambah iterasi sebanyak 1
        
    return x, iterasi # Nilai yang direturn fungsi ini adalah x(k) dan jumlah iterasi

print('Aproksimasi solusi SPL tersebut dengan metode Jacobi adalah:',Jacobi(aug_matrix,initial_solution,tolerance)) # Panggil fungsi Jacobi dengan parameter yang diisi

Masukkan augmented matrix dari SPL yang akan diselesaikan: [[1,4,1,2,8],[1,4,1,1,7],[1,2,3,2,8],[1,1,7,4,14]]
Masukkan tebakan awal solusi: [1,1,1,1]
Masukkan toleransi: 10**-4
[[ 1  4  1  2  8]
 [ 1  4  1  1  7]
 [ 1  2  3  2  8]
 [ 1  1  7  4 14]]


C:\Users\Kamal\AppData\Local\Temp/ipykernel_4228/1025722994.py:31: RuntimeWarning: overflow encountered in double_scalars
  b-=matrix[i,j]*x0[j]
C:\Users\Kamal\AppData\Local\Temp/ipykernel_4228/1025722994.py:31: RuntimeWarning: overflow encountered in multiply
  b-=matrix[i,j]*x0[j]


KeyboardInterrupt: 

Program di atas adalah program untuk mencari aproksimasi SPL dengan metode Jacobi.

Idenya adalah mencari solusi SPL dari input matriks, tebakan awal, dan toleransi dari user dengan metode Gauss-Seidel serta menampilkan hasil dan jumlah iterasi.

Algoritmanya:
1. *Import* module `numpy`, lalu panggil sebagai `np`
1. *Import* `tabulate` dari module `tabulate`
1. Simpan *input* matriks dari *user* dan simpan dalam variabel `aug_matrix`
1. Simpan *input* tebakan awal dari *user* dan simpan dalam variabel `initial_solution`
1. Simpan *input* batas toleransi dari *user* dan simpan dalam variabel `tolerance`
1. Cetak matriks `aug_matrix`
1. Buat fungsi metode Jacobi dengan *argument* `matrix`, `initial`, dan `tol`. Berdasarkan Modul 6 Praktikum Metnum, program metode Jacobi akan berjalan sesuai dengan ketentuan tersebut, kecuali:
    1. Inisialisasi jumlah iterasi dengan nilai `1` ke dalam variabel `iterasi`
    1. Tambah nilai variabel `iterasi` sebanyak `1` pada tiap pengulangan while loop
    1. *Return*-kan nilai `x` dan `iterasi` setelah fungsi selesai
1. Panggil fungsi Jacobi dengan argumen `aug_matrix`, `initial_solution`, dan `tolerance`. Kemudian, cetak hasil dan jumlah iterasi dari variabel `x` dan `iterasi` yang sudah di *return* dari fungsi.

In [6]:
import numpy as np # Import numpy untuk array
from tabulate import tabulate # Import tabulate untuk tabel

aug_matrix = np.array(eval(input('Masukkan augmented matrix dari SPL yang akan diselesaikan: '))) # Simpan matriks yang akan diselesaikan
initial_solution = np.array(eval(input('Masukkan tebakan awal solusi: ')),dtype=float) # Simpan tebakan awal solusi dari user (x0)
tolerance = eval(input('Masukkan toleransi: ')) # Simpan batas toleransi yang diinginkan

print(aug_matrix) # Cetak matriks

def GaussSeidel(matrix,initial,tol):
    n = np.shape(matrix)[0] #ukuran matriks/banyak variabel
    x0 = initial #definisikan x0 sebagai tebakan awal
    x = x0.copy() #definisikan x = x0 (gunakan .copy() agar jika x0 berubah, x tidak berubah)
    
    for i in range(n):
        b = matrix[i,n] # Konstanta pada SPL
        for j in range(n): # Metode Gauss-Seidel
            if j!=i:
                if j<i:
                    b-=matrix[i,j]*x[j]
                else:
                    b-=matrix[i,j]*x0[j]
        x[i]=b/matrix[i,i] # Hasil aproksimasi Gauss-Seidel

    iterasi = 1 # Inisiasi nilai jumlah iterasi
    
    while np.linalg.norm(x-x0,np.inf)>=tol: # Pengecekan kriteria penyetop menggunakan error mutlak
        x0 = x # Definisikan x0 sebagai x(k-1)
        x = x0.copy() # Definisikan x sebagai x(k), namun awalnya samakan dengan x0
        for i in range(n): # Metode Gauss-Seidel
            b = matrix[i,n] # Konstanta pada SPL
            for j in range(n):
                if j!=i:
                    if j<i:
                        b-=matrix[i,j]*x[j]
                    else:
                        b-=matrix[i,j]*x0[j]
            x[i]=b/matrix[i,i] # Hasil aproksimasi metode Gauss-Seidel
        
        iterasi += 1 # Tambah iterasi sebanyak 1
        
    return x, iterasi # Nilai yang direturn fungsi ini adalah x(k) dan jumlah iterasi

print('Aproksimasi solusi SPL tersebut dengan metode Gauss-Seidel adalah:',GaussSeidel(aug_matrix,initial_solution,tolerance)) # Panggil fungsi Gauss-Seidel dengan parameter yang diisi

Masukkan augmented matrix dari SPL yang akan diselesaikan: [[1,4,1,2,8],[1,4,1,1,7],[1,2,3,2,8],[1,1,7,4,14]]
Masukkan tebakan awal solusi: [1,1,1,1]
Masukkan toleransi: 10**-4
[[ 1  4  1  2  8]
 [ 1  4  1  1  7]
 [ 1  2  3  2  8]
 [ 1  1  7  4 14]]
Aproksimasi solusi SPL tersebut dengan metode Gauss-Seidel adalah: (array([inf, nan, nan, nan]), 27617)


C:\Users\Kamal\AppData\Local\Temp/ipykernel_4228/537030765.py:35: RuntimeWarning: overflow encountered in multiply
  b-=matrix[i,j]*x[j]
C:\Users\Kamal\AppData\Local\Temp/ipykernel_4228/537030765.py:37: RuntimeWarning: invalid value encountered in double_scalars
  b-=matrix[i,j]*x0[j]


Program di atas adalah program untuk mencari aproksimasi SPL dengan metode Gauss-Seidel.

Idenya adalah mencari solusi SPL dari input matriks, tebakan awal, dan toleransi dari user dengan metode Gauss-Seidel serta menampilkan hasil dan jumlah iterasinya.

Algoritmanya:
1. *Import* module `numpy`, lalu panggil sebagai `np`
1. *Import* `tabulate` dari module `tabulate`
1. Simpan *input* matriks dari *user* dan simpan dalam variabel `aug_matrix`
1. Simpan *input* tebakan awal dari *user* dan simpan dalam variabel `initial_solution`
1. Simpan *input* batas toleransi dari *user* dan simpan dalam variabel `tolerance`
1. Cetak matriks `aug_matrix`
1. Buat fungsi metode Gauss-Seidel dengan *argument* `matrix`, `initial`, dan `tol`. Berdasarkan Modul 6 Praktikum Metode Numerik, program metode Gauss-Seidel akan berjalan sesuai dengan ketentuan tersebut, kecuali:
    1. Inisialisasi jumlah iterasi dengan nilai `1` ke dalam variabel `iterasi`
    1. Tambah nilai variabel `iterasi` sebanyak `1` pada tiap pengulangan while loop
    1. *Return*-kan nilai `x` dan `iterasi` setelah fungsi selesai
1. Panggil fungsi Gauss-Seidel dengan argumen `aug_matrix`, `initial_solution`, dan `tolerance`. Kemudian, cetak hasil dan jumlah iterasi dari variabel `x` dan `iterasi` yang sudah di *return* dari fungsi.

### Jawaban

2. Pada metode Jacobi, program tidak menampilkan hasil iterasi karena tidak berhasil mendapatkan solusi dari matriks diperbesar yang diberikan, terjadi *overflow* ketika program dijalankan. Dapat dipastikan bahwa jumlah iterasinya terlalu banyak hingga program tidak dapat memberikan output.

    Pada metode Gauss-Seidel, solusi SPL tersebut yang diperoleh bernilai `[inf, nan, nan, nan]` dengan keterangan *infinity* (`inf`) dan *not a number* (`nan`). Dapat dipastikan bahwa solusi tidak akan konvergen atau solusi divergen. Didapatkan juga jumlah iterasinya, yakni sebanyak 27617 hingga program berhenti.

    Dapat disimpulkan bahwa metode Gauss-Seidel lebih baik dalam mengaproksimasi SPL dibandingkan dengan metode Jacobi, seperti pada soal nomor 1.